utils/test_utils.py

```diff
@@ -0,0 +1,259 @@
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_5():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=5,
+        existing_sd_files=[],
+        expected_sd_files=[
+            '.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc_0():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=0,
+        existing_sd_files=[],
+        expected_sd_files=[
+            '.gcode',
+            '1.gcode',
+            '2.gcode',
+            '3.gcode',
+            '4.gcode',
+            '5.gcode',
+            '6.gcode',
+            '7.gcode',
+            '8.gcode',
+            '9.gcode',
+            '10.gcode',
+            '11.gcode']
+    )
+
+def test_generate_new_filename_abc1_30_existing_abc1():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc1.gcode',
+        max_filename=30,
+        existing_sd_files=['abc1.gcode'],
+        expected_sd_files=[
+            'abc1.gcode',
+            'abc11.gcode',
+            'abc12.gcode',
+            'abc13.gcode',
+            'abc14.gcode',
+            'abc15.gcode',
+            'abc16.gcode',
+            'abc17.gcode',
+            'abc18.gcode',
+            'abc19.gcode',
+            'abc110.gcode',
+            'abc111.gcode',
+            'abc112.gcode']
+    )
+
+def test_generate_new_filename_abc_30_existing_abc5():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc.gcode',
+        max_filename=30,
+        existing_sd_files=['abc5.gcode'],
+        expected_sd_files=[
+            'abc5.gcode',
+            'abc.gcode',
+            'abc1.gcode',
+            'abc2.gcode',
+            'abc3.gcode',
+            'abc4.gcode',
+            'abc6.gcode',
+            'abc7.gcode',
+            'abc8.gcode',
+            'abc9.gcode',
+            'abc10.gcode',
+            'abc11.gcode',
+            'abc12.gcode']
+    )
+
+def test_generate_new_filename_abc_no_file_extension():
+    simulate_printing_12_times_and_validate_result(
+        filename='abc',
+        max_filename=30,
+        existing_sd_files=[],
+        expected_sd_files=[
+            'abc',
+            'abc1',
+            'abc2',
+            'abc3',
+            'abc4',
+            'abc5',
+            'abc6',
+            'abc7',
+            'abc8',
+            'abc9',
+            'abc10',
+            'abc11']
+    )
+
+def test_generate_new_filename_hardname():
+    simulate_printing_12_times_and_validate_result(
+        filename='aBc$.de..ext',
+        max_filename=30,
+        existing_sd_files=[],
+        expected_sd_files=[
+            'aBc$.de..ext',
```

Comment:
2 dots, should not be expected here, because it's incorrect file name extension

Reply:
this is also valid a case
![image](https://user-images.githubusercontent.com/5782954/149216925-5e41ee95-733f-4468-9f57-b13b1482f2a4.png)

Reply:
ok, then. It's a trifle anyway, let's keep it as it is.